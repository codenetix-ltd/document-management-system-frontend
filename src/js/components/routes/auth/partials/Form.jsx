import React, { Component } from 'react';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';

import ls from 'Services/SecureLS';

@autobind
export default class AuthForm extends Component {
  static defaultProps = {
    submitButtonText: 'Login',
    validate: () => false
  };

  static propTypes = {
    onSubmit: PropTypes.func.isRequired,
    submitButtonText: PropTypes.string,
    validate: PropTypes.func
  };

  constructor(props) {
    super(props);
    const { email, password } = ls.get('c_cache');
    this.state = {
      email: email || '',
      password: password || '',
      remember: false
    };
  }

  /**
   * Handles submit event.
   * @param e
   */
  onSubmit(e) {
    e.preventDefault();
    const { email, password, remember } = this.state;
    if (remember) ls.set('c_cache', { email, password });
    this.props.onSubmit({ email, password });
  }

  /**
   * Universal handler for multiple input fields
   * @param event
   */
  handleChange(event) {
    const { target } = event;
    const { value, name } = target;
    this.setState({
      [name]: value
    });
  }

  toggleCheckbox() {
    this.setState({ remember: !this.state.remember });
  }

  render() {
    return (
      <form onSubmit={this.onSubmit}>
        <div className="form-group has-feedback">
          <input
            type="email"
            className="form-control"
            placeholder="Email"
            name="email"
            value={this.state.email}
            onChange={this.handleChange}
          />
          <span className="form-control-feedback">
            <i className="fa fa-envelope" />
          </span>
        </div>
        <div className="form-group has-feedback">
          <input
            type="password"
            className="form-control"
            placeholder="Password"
            name="password"
            value={this.state.password}
            onChange={this.handleChange}
          />
          <span className="form-control-feedback">
            <i className="fa fa-lock" />
          </span>
        </div>
        <div className="row">
          <div className="col-xs-8">
            <div className="checkbox">
              <label>
                <div className="icheckbox_square-blue">
                  <input
                    type="checkbox"
                    checked={this.state.remember}
                    onChange={this.toggleCheckbox}
                  />
                </div>
                Remember Me
              </label>
            </div>
          </div>
          <div className="col-xs-4">
            <button
              type="submit"
              className="btn btn-primary btn-block btn-flat"
              disabled={this.props.validate(this.state)}
            >
              {this.props.submitButtonText}
            </button>
          </div>
        </div>
      </form>
    );
  }
}
