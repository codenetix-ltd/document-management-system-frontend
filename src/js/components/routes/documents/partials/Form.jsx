import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';
import { Async as Select } from 'react-select';
import { If, Then } from 'qc-react-conditionals/lib';
import pick from 'lodash/pick';

import FileUpload from 'Components/common/FileUpload';
import FormError from 'Components/common/FormError';

import {
  $$documentUpdate,
  $$documentReset
} from 'Store/thunks/documents';
import { $$templatesFetch } from 'Store/thunks/templates';
import { $$labelsFetch } from 'Store/thunks/labels';
import { $setAttributeValues } from 'Store/actions';
import { $$errorsReset, $$errorsUpdate } from 'Store/thunks/errors';

@autobind
export class DocumentForm extends Component {
  static propTypes = {
    profile: PropTypes.any.isRequired,
    document: PropTypes.any.isRequired,
    templates: PropTypes.any.isRequired,
    labels: PropTypes.any.isRequired,
    dispatch: PropTypes.func.isRequired
  };

  componentWillUnmount() {
    const { dispatch } = this.props;
    $$errorsReset(dispatch);
    $$documentReset(dispatch);
  }

  getLabelOptions(input, callback) {
    $$labelsFetch(this.props.dispatch, 1, () => {
      const { list } = this.props.labels;
      callback(null, { options: list, complete: true });
    });
  }

  getTemplateOptions(input, callback) {
    $$templatesFetch(this.props.dispatch, 1, () => {
      const { list } = this.props.templates;
      callback(null, { options: list, complete: true });
    });
  }

  getEmptyOptions(input, callback) {
    callback(null, { options: [] });
  }

  handleChange({ target }) {
    const { dispatch } = this.props;
    const { name, value, dataset } = target;
    $$documentUpdate(dispatch, {
      [name]: value
    });
    if (dataset.field) {
      $$errorsUpdate(dispatch, {
        [dataset.field]: ''
      });
    }
  }

  handleLabelsSelect(value) {
    $$documentUpdate(this.props.dispatch, {
      labels: value
    });
  }

  handleTemplateSelect(value) {
    const { dispatch } = this.props;
    dispatch($setAttributeValues([]));
    $$documentUpdate(dispatch, { template: value });
    $$errorsUpdate(dispatch, { 'actualVersion.templateId': '' });
  }

  handleFileUpload(fileIds) {
    $$documentUpdate(this.props.dispatch, { fileIds });
  }

  render() {
    const { document, document: { actualVersion }, profile } = this.props;
    return (
      <form className="form-horizontal">
        <div className="box-body">
          <div className="form-group">
            <label htmlFor="name" className="col-sm-2 control-label">Name</label>
            <div className="col-sm-6">
              <input
                type="text"
                name="name"
                className="form-control"
                placeholder="Document name"
                value={actualVersion.name}
                onChange={this.handleChange}
                data-field="actualVersion.name"
              />
              <FormError field="actualVersion.name" />
            </div>
          </div>
          <If is={document.version}>
            <Then>
              <div className="form-group">
                <label htmlFor="version" className="col-sm-2 control-label">Actual version</label>
                <div className="col-sm-6">
                  <Select
                    name="version"
                    value={{ id: 1, name: document.version }}
                    disabled
                    labelKey="name"
                    valueKey="id"
                    loadOptions={this.getEmptyOptions}
                  />
                </div>
              </div>
            </Then>
          </If>
          <div className="form-group">
            <label htmlFor="template" className="col-sm-2 control-label">Template</label>
            <div className="col-sm-6">
              <Select
                autoload
                name="template"
                value={actualVersion.template}
                onChange={this.handleTemplateSelect}
                loadOptions={this.getTemplateOptions}
                labelKey="name"
                valueKey="id"
              />
              <FormError field="actualVersion.templateId" />
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="comment" className="col-sm-2 control-label">Comment</label>
            <div className="col-sm-6">
              <textarea
                name="comment"
                className="form-control"
                placeholder="Document comment"
                value={actualVersion.comment}
                onChange={this.handleChange}
              />
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="labels" className="col-sm-2 control-label">Labels</label>
            <div className="col-sm-6">
              <Select
                multi
                autoload
                name="labels"
                value={actualVersion.labels}
                onChange={this.handleLabelsSelect}
                loadOptions={this.getLabelOptions}
                labelKey="name"
                valueKey="id"
              />
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="owner" className="col-sm-2 control-label">Owner</label>
            <div className="col-sm-6">
              <Select
                name="owner"
                value={profile}
                disabled
                labelKey="fullName"
                valueKey="id"
                loadOptions={this.getEmptyOptions}
              />
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="file" className="col-sm-2 control-label">Files</label>
            <div className="col-sm-6">
              <FileUpload multiple onSuccess={this.handleFileUpload} />
            </div>
          </div>
        </div>
      </form>
    );
  }
}

const mapStateToProps = (state) => pick(state, ['document', 'templates', 'labels', 'profile']);

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(DocumentForm);
