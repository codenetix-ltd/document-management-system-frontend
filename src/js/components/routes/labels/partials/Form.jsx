import React, { Component } from 'react';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';

@autobind
export default class LabelForm extends Component {
  static defaultProps = {
    label: null,
    submitButtonText: 'Create',
    validate: () => false
  };

  static propTypes = {
    label: PropTypes.any,
    onSubmit: PropTypes.func.isRequired,
    submitButtonText: PropTypes.string,
    validate: PropTypes.func
  };

  constructor(props) {
    super(props);
    this.state = {
      name: ''
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.label !== this.props.label) {
      this.setState({ ...this.state, ...nextProps.label });
    }
  }

  /**
   * Handles submit event.
   * @param event
   */
  onSubmit(event) {
    event.preventDefault();
    const formData = { ...this.state };
    this.props.onSubmit(formData);
  }

  /**
   * Universal handler for multiple input fields
   * @param event
   */
  handleChange(event) {
    const { target } = event;
    const { value, name } = target;
    this.setState({
      [name]: value
    });
  }

  render() {
    return (
      <form className="form-horizontal" onSubmit={this.onSubmit}>
        <div className="box-body">
          <div className="form-group">
            <label htmlFor="name" className="col-sm-2 control-label">Name</label>
            <div className="col-sm-6">
              <input
                className="form-control"
                placeholder="Name"
                type="text"
                name="name"
                value={this.state.name}
                onChange={this.handleChange}
              />
            </div>
          </div>
        </div>
        <div className="box-footer">
          <button
            className="btn btn-success"
            type="submit"
            disabled={this.props.validate(this.state)}
          >
            {this.props.submitButtonText}
          </button>
        </div>
      </form>
    );
  }
}
