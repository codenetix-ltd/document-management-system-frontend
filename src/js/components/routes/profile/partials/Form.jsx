import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';

import FileUpload from 'Components/common/FileUpload';

import { $$profileUpdate } from 'Store/thunks/profile';

@autobind
export class ProfileForm extends Component {
  static propTypes = {
    profile: PropTypes.any.isRequired,
    dispatch: PropTypes.func.isRequired,
    onSubmit: PropTypes.func.isRequired
  };

  /**
   * Handles submit event.
   * @param e - Event
   */
  onSubmit(e) {
    e.preventDefault();
    this.props.onSubmit();
  }

  /**
   * Universal handler for multiple input fields
   * @param event
   */
  handleChange({ target: { value, name } }) {
    $$profileUpdate(this.props.dispatch, {
      [name]: value
    });
  }

  /**
   * Handler for file upload component
   * @param fileIds
   */
  handleAvatarUpload(fileIds) {
    $$profileUpdate(this.props.dispatch, {
      avatarId: fileIds[0]
    });
  }

  render() {
    const { profile } = this.props;
    return (
      <form className="form-horizontal" onSubmit={this.onSubmit}>
        <div className="box-body">
          <div className="form-group">
            <label htmlFor="full_name" className="col-sm-2 control-label">Full name</label>
            <div className="col-sm-6">
              <input
                id="full_name"
                className="form-control"
                placeholder="Full Name"
                type="text"
                name="fullName"
                value={profile.fullName}
                onChange={this.handleChange}
              />
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="email" className="col-sm-2 control-label">Email</label>
            <div className="col-sm-6">
              <input
                id="email"
                className="form-control"
                placeholder="Email"
                type="email"
                name="email"
                value={profile.email}
                onChange={this.handleChange}
              />
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="file" className="col-sm-2 control-label">Avatar</label>
            <div className="col-sm-6">
              <FileUpload onSuccess={this.handleAvatarUpload} />
            </div>
          </div>
        </div>
        <div className="box-footer">
          <button className="btn btn-success" type="submit">
            Update
          </button>
        </div>
      </form>
    );
  }
}

const mapStateToProps = ({ profile }) => ({ profile });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(ProfileForm);
