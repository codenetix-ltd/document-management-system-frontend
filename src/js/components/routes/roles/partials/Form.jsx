import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';
import Select from 'react-select';
import { If, Then } from 'qc-react-conditionals/lib';

import { $$permissionGroupsFetch } from 'Store/thunks/roles';
import { $$templatesFetch } from 'Store/thunks/templates';

@autobind
export class RoleForm extends Component {
  static defaultProps = {
    role: null
  };

  static propTypes = {
    role: PropTypes.any,
    onSubmit: PropTypes.func.isRequired,
    dispatch: PropTypes.func.isRequired,
    loading: PropTypes.bool.isRequired,
    templates: PropTypes.any.isRequired,
    permissionGroups: PropTypes.array.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      name: '',
      permissionValues: [],
      templateIds: []
    };
  }

  componentDidMount() {
    const { dispatch } = this.props;
    $$permissionGroupsFetch(dispatch, () => {
      $$templatesFetch(dispatch, 0);
    });
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.role !== this.props.role) {
      this.setState({ ...this.state, ...nextProps.role });
    }
  }

  /**
   * Handles submit event.
   * @param event
   */
  onSubmit(event) {
    event.preventDefault();
    const formData = { ...this.state };
    this.props.onSubmit(formData);
  }

  /**
   * Universal handler for multiple input fields
   * @param event
   */
  handleChange(event) {
    const { target } = event;
    const { value, name } = target;
    this.setState({ [name]: value });
  }

  /**
   * Special handler for react-select's Async component
   * @param value
   */
  handleTemplatesSelect(value) {
    this.setState({ templatesIds: value });
  }

  render() {
    const { name, templateIds } = this.state;
    const {
      role,
      loading,
      templates,
      permissionGroups
    } = this.props;
    const userTemplates = templates.list.filter(tpl => templateIds.includes(tpl));
    return (
      <form onSubmit={this.onSubmit}>
        <h3>General information</h3>
        <hr />
        <div className="row">
          <div className="col-lg-5">
            <div className="form-group">
              <label htmlFor="name" className="control-label">Name</label>
              <input
                type="text"
                name="name"
                className="form-control"
                placeholder="Role name"
                value={name}
                onChange={this.handleChange}
              />
            </div>
            <div className="form-group">
              <label htmlFor="template" className="control-label">Attached templates</label>
              <Select
                multi
                isLoading={loading}
                name="templates"
                value={userTemplates}
                onChange={this.handleTemplatesSelect}
                options={templates.list}
                labelKey="name"
                valueKey="id"
              />
            </div>
          </div>
        </div>
        <If is={role}>
          <Then>
            <div className="row">
              <div className="col-lg-8">
                <h3>Permissions</h3>
                <hr />
                <div id="permission-list-wrapper">
                  {
                    permissionGroups.map(group => (
                      <div key={group.id} className="permissionGroup">
                        <h4>{group.label}</h4>
                        <table className="table table-bordered table-permissions">
                          <thead>
                            <tr>
                              <th>Permission</th>
                              <th>Access type</th>
                              <th className="qualifier" />
                            </tr>
                          </thead>
                          <tbody>
                            {
                              group['permissions'].map(permission => (
                                <tr key={permission.id}>
                                  <td>{permission.label}</td>
                                  <td>
                                    <select name="access_type[1]" className="form-control input-sm">
                                      {
                                        permission['accessTypes'].map(accessType => (
                                          <option key={accessType.id} value={accessType.id}>
                                            {accessType.label}
                                          </option>
                                        ))
                                      }
                                    </select>
                                  </td>
                                  <td className="qualifier">
                                    <div className="qualifier-wrapper">
                                      {
                                        group['qualifiers'].map(qualifier => (
                                          <div key={qualifier.id}>
                                            <span>{qualifier.label}</span>
                                            <select name="qualifier[1][2]" className="form-control input-sm">
                                              {
                                                qualifier['accessTypes'].map(accessType => (
                                                  <option key={accessType.id} value={accessType.id}>
                                                    {accessType.label}
                                                  </option>
                                                ))
                                              }
                                            </select>
                                          </div>
                                        ))
                                      }
                                    </div>
                                  </td>
                                </tr>
                              ))
                            }
                          </tbody>
                        </table>
                      </div>
                    ))
                  }
                </div>
              </div>
            </div>
          </Then>
        </If>
        <div>
          <button type="submit" className="btn btn-success">
            { role ? 'Update' : 'Create' }
          </button>
        </div>
      </form>
    );
  }
}

const mapStateToProps = ({ permissionGroups, templates, loading }) => ({ permissionGroups, templates, loading });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(RoleForm);
