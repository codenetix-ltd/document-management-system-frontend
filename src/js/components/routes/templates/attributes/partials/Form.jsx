import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import autobind from 'autobind-decorator';
import { If, Then } from 'qc-react-conditionals/lib';
import Select from 'react-select';

import FormError from 'Components/common/FormError';
import { DataInput as Input } from 'Components/common/dataControls';
import AttributesTable from 'Routes/templates/attributes/partials/Table';

import {
  $$attributeUpdate,
  $$attributeReset
} from 'Store/thunks/attributes';

import {
  $$errorsReset,
  $$errorsUpdate
} from 'Store/thunks/errors';

@autobind
export class AttributeForm extends Component {
  static defaultProps = {
    submitButtonText: 'Create'
  };

  static propTypes = {
    types: PropTypes.array.isRequired,
    attribute: PropTypes.any.isRequired,
    dispatch: PropTypes.any.isRequired,
    onSubmit: PropTypes.func.isRequired,
    submitButtonText: PropTypes.string
  };

  componentWillUnmount() {
    const { dispatch } = this.props;
    $$attributeReset(dispatch);
    $$errorsReset(dispatch);
  }

  onSubmit(e) {
    e.preventDefault();
    const { attribute } = this.props;
    const { name, type, data } = { ...attribute };
    this.props.onSubmit({
      name,
      data,
      typeId: type && type.id
    });
  }

  handleChange(e) {
    const { value } = e.target;
    const { dispatch } = this.props;
    $$attributeUpdate(dispatch, { name: value });
    $$errorsUpdate(dispatch, { name: '' });
  }

  handleSelectChange(type) {
    const { dispatch } = this.props;
    $$attributeUpdate(dispatch, { type });
    $$errorsUpdate(dispatch, { typeId: '' });
  }

  render() {
    const { attribute, types } = this.props;
    return (
      <form className="form-horizontal" onSubmit={this.onSubmit}>
        <div className="box-body">
          <div className="form-group">
            <label htmlFor="name" className="col-sm-2 control-label">Name</label>
            <div className="col-sm-6">
              <Input
                debounce
                id="name"
                name="name"
                type="text"
                cls="form-control"
                value={attribute.name}
                placeholder="Attribute name"
                onChange={this.handleChange}
              />
              <FormError field="name" />
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="type" className="col-sm-2 control-label">Type</label>
            <div className="col-sm-6">
              <Select
                name="type"
                value={attribute.type}
                onChange={this.handleSelectChange}
                options={types}
                valueKey="id"
                labelKey="name"
              />
              <FormError field="typeId" />
            </div>
          </div>
          <If is={attribute.type && attribute.type.name === 'Table'}>
            <Then>
              <div className="form-group form-group-table">
                <label className="col-sm-2 control-label">Table definition</label>
                <div className="col-sm-10">
                  <AttributesTable />
                </div>
              </div>
            </Then>
          </If>
        </div>
        <div className="box-footer">
          <button className="btn btn-success" type="submit">
            {this.props.submitButtonText}
          </button>
        </div>
      </form>
    );
  }
}

const mapStateToProps = ({ attribute, types }) => ({ attribute, types });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(AttributeForm);
