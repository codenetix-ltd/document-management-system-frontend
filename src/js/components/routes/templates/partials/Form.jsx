import axios from 'axios';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';
import { Link, Redirect } from 'react-router-dom';
import { If, Then } from 'qc-react-conditionals/lib';

import AttributesList from 'Routes/templates/partials/AttributesListSortable';

import {
  $$templateFetch,
  $$templateUpdate,
  $$templateReset
} from 'Store/thunks/templates';

import { API } from 'Config';

@autobind
export class TemplateForm extends Component {
  static defaultProps = {
    match: {
      params: {}
    },
    submitButtonText: 'Create',
    validate: () => false
  };

  static propTypes = {
    match: PropTypes.any,
    template: PropTypes.any.isRequired,
    attributes: PropTypes.any.isRequired,
    dispatch: PropTypes.func.isRequired,
    submitButtonText: PropTypes.string,
    validate: PropTypes.func
  };

  constructor(props) {
    super(props);
    this.state = { newTemplateID: false };
  }

  componentDidMount() {
    const { dispatch, match } = this.props;
    const { templateID } = match.params;
    if (templateID) {
      $$templateFetch(dispatch, templateID);
    }
  }

  componentWillUnmount() {
    $$templateReset(this.props.dispatch);
  }

  onSubmit(event) {
    event.preventDefault();
    const { templateID } = this.props.match.params;
    const { name } = this.props.template;
    const { list } = this.props.attributes;
    const formData = {
      name,
      orderOfAttributes: list.map(item => item.id)
    };
    if (templateID) {
      axios.put(`${API.templates}/${templateID}`, formData).catch(err => {
        throw err;
      });
    } else {
      axios.post(API.templates, formData).then(({ data }) => {
        if (!data.id) throw new Error('response has no id field');
        this.setState({
          newTemplateID: data.id
        });
      }).catch(err => {
        throw err;
      });
    }
  }

  handleChange(event) {
    const { target } = event;
    const { value, name } = target;
    const { dispatch } = this.props;
    $$templateUpdate(dispatch, {
      [name]: value
    });
  }

  render() {
    const { templateID } = this.props.match.params;
    const { template } = this.props;

    if (this.state.newTemplateID) {
      return (
        <Redirect to={`/templates/${this.state.newTemplateID}`} />
      );
    }

    return (
      <form className="form-horizontal" onSubmit={this.onSubmit}>
        <div className="box-body">
          <div className="form-group">
            <label htmlFor="name" className="col-sm-2 control-label">Name</label>
            <div className="col-sm-6">
              <input
                id="name"
                className="form-control"
                placeholder="Template name"
                type="text"
                name="name"
                value={template.name}
                onChange={this.handleChange}
              />
            </div>
          </div>
          <If is={templateID}>
            <Then>
              <div className="form-group">
                <label className="col-sm-2 control-label">Attributes</label>
                <div className="col-sm-6">
                  <AttributesList templateID={templateID} prompt={this.prompt} />
                  <Link className="btn btn-primary pull-right btn-sm" to={`/templates/${templateID}/attributes`}>
                    <i className="fa fa-plus" /> Add attribute
                  </Link>
                </div>
              </div>
            </Then>
          </If>
        </div>
        <div className="box-footer">
          <button
            className="btn btn-success"
            type="submit"
            disabled={this.props.validate(template)}
          >
            {this.props.submitButtonText}
          </button>
        </div>
      </form>
    );
  }
}

const mapStateToProps = ({ template, attributes }) => ({ template, attributes });

const mapDispatchToProps = (dispatch) => ({ dispatch });

export default connect(mapStateToProps, mapDispatchToProps)(TemplateForm);
