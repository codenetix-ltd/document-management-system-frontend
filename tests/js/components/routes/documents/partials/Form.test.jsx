import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { DocumentForm } from 'Routes/documents/partials/Form';

import {
  $$documentUpdate,
  $$documentReset
} from 'Store/thunks/documents';
import { $$labelsFetch } from 'Store/thunks/labels';
import { $$templatesFetch } from 'Store/thunks/templates';

jest.mock('Store/thunks/documents', () => ({
  $$documentReset: jest.fn(),
  $$documentUpdate: jest.fn()
}));

jest.mock('Store/thunks/labels', () => ({
  $$labelsFetch: jest.fn()
}));

jest.mock('Store/thunks/templates', () => ({
  $$templatesFetch: jest.fn()
}));

describe('Documents Form', () => {
  let wrapper;
  let props;
  let instance;

  beforeEach(() => {
    props = {
      dispatch: () => {},
      document: {
        actualVersion: {
          template: {
            attributes: []
          }
        }
      },
      profile: {},
      templates: {
        list: []
      },
      labels: {
        list: []
      }
    };
    wrapper = shallow(<DocumentForm {...props} />);
    instance = wrapper.instance();
  });

  it('should render correctly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should call $$documentReset on componentWillUnmount', () => {
    expect($$documentReset).not.toHaveBeenCalled();
    wrapper.unmount();
    expect($$documentReset).toHaveBeenCalledTimes(1);
  });

  it('should call $$labelsFetch when calling getLabelOptions', () => {
    expect($$labelsFetch).not.toHaveBeenCalled();
    instance.getLabelOptions('', () => {});
    expect($$labelsFetch).toHaveBeenCalledTimes(1);
  });

  it('should call $$templatesFetch when calling getLabelOptions', () => {
    expect($$templatesFetch).not.toHaveBeenCalled();
    instance.getTemplateOptions('', () => {});
    expect($$templatesFetch).toHaveBeenCalledTimes(1);
  });

  it('calls getEmptyOptions method', () => {
    const callback = jest.fn();
    expect(callback).not.toHaveBeenCalled();
    instance.getEmptyOptions('', callback);
    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toBeCalledWith(null, { options: [] });
  });

  it('should call $$documentUpdate in handleChange', () => {
    expect($$documentUpdate).not.toHaveBeenCalled();
    instance.handleChange({
      target: {
        name: 'lorem',
        value: 'ipsum',
        dataset: {
          field: ''
        }
      }
    });
    expect($$documentUpdate).toHaveBeenCalledTimes(1);
    expect($$documentUpdate).toBeCalledWith(props.dispatch, { lorem: 'ipsum' });
  });

  it('should call $$documentUpdate in handleLabelsSelect', () => {
    const value = 'test';
    expect($$documentUpdate).not.toHaveBeenCalled();
    instance.handleLabelsSelect(value);
    expect($$documentUpdate).toHaveBeenCalledTimes(1);
    expect($$documentUpdate).toBeCalledWith(props.dispatch, { labels: value });
  });

  it('should call $$documentUpdate in handleTemplateSelect', () => {
    const value = 'test';
    expect($$documentUpdate).not.toHaveBeenCalled();
    instance.handleTemplateSelect(value);
    expect($$documentUpdate).toHaveBeenCalledTimes(1);
    expect($$documentUpdate).toBeCalledWith(props.dispatch, { template: value });
  });

  it('should call $$documentUpdate in handleFileUpload', () => {
    const fileIds = [1, 2, 3];
    expect($$documentUpdate).not.toHaveBeenCalled();
    instance.handleFileUpload(fileIds);
    expect($$documentUpdate).toHaveBeenCalledTimes(1);
    expect($$documentUpdate).toBeCalledWith(props.dispatch, { fileIds });
  });
});
