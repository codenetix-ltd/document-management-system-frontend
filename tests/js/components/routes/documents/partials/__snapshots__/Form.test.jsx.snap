// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Documents Form should render correctly 1`] = `
<form
  className="form-horizontal"
>
  <div
    className="box-body"
  >
    <div
      className="form-group"
    >
      <label
        className="col-sm-2 control-label"
        htmlFor="name"
      >
        Name
      </label>
      <div
        className="col-sm-6"
      >
        <input
          className="form-control"
          data-field="actualVersion.name"
          name="name"
          onChange={[Function]}
          placeholder="Document name"
          type="text"
        />
        <Connect(FormError)
          field="actualVersion.name"
        />
      </div>
    </div>
    <If>
      <Then>
        <div
          className="form-group"
        >
          <label
            className="col-sm-2 control-label"
            htmlFor="version"
          >
            Actual version
          </label>
          <div
            className="col-sm-6"
          >
            <Async
              autoload={true}
              cache={Object {}}
              disabled={true}
              ignoreAccents={true}
              ignoreCase={true}
              labelKey="name"
              loadOptions={[Function]}
              loadingPlaceholder="Loading..."
              name="version"
              options={Array []}
              searchPromptText="Type to search"
              value={
                Object {
                  "id": 1,
                  "name": undefined,
                }
              }
              valueKey="id"
            />
          </div>
        </div>
      </Then>
    </If>
    <div
      className="form-group"
    >
      <label
        className="col-sm-2 control-label"
        htmlFor="template"
      >
        Template
      </label>
      <div
        className="col-sm-6"
      >
        <Async
          autoload={true}
          cache={Object {}}
          ignoreAccents={true}
          ignoreCase={true}
          labelKey="name"
          loadOptions={[Function]}
          loadingPlaceholder="Loading..."
          name="template"
          onChange={[Function]}
          options={Array []}
          searchPromptText="Type to search"
          value={
            Object {
              "attributes": Array [],
            }
          }
          valueKey="id"
        />
        <Connect(FormError)
          field="actualVersion.templateId"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        className="col-sm-2 control-label"
        htmlFor="comment"
      >
        Comment
      </label>
      <div
        className="col-sm-6"
      >
        <textarea
          className="form-control"
          name="comment"
          onChange={[Function]}
          placeholder="Document comment"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        className="col-sm-2 control-label"
        htmlFor="labels"
      >
        Labels
      </label>
      <div
        className="col-sm-6"
      >
        <Async
          autoload={true}
          cache={Object {}}
          ignoreAccents={true}
          ignoreCase={true}
          labelKey="name"
          loadOptions={[Function]}
          loadingPlaceholder="Loading..."
          multi={true}
          name="labels"
          onChange={[Function]}
          options={Array []}
          searchPromptText="Type to search"
          valueKey="id"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        className="col-sm-2 control-label"
        htmlFor="owner"
      >
        Owner
      </label>
      <div
        className="col-sm-6"
      >
        <Async
          autoload={true}
          cache={Object {}}
          disabled={true}
          ignoreAccents={true}
          ignoreCase={true}
          labelKey="fullName"
          loadOptions={[Function]}
          loadingPlaceholder="Loading..."
          name="owner"
          options={Array []}
          searchPromptText="Type to search"
          value={Object {}}
          valueKey="id"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        className="col-sm-2 control-label"
        htmlFor="file"
      >
        Files
      </label>
      <div
        className="col-sm-6"
      >
        <FileUpload
          multiple={true}
          onFailure={[Function]}
          onSuccess={[Function]}
        />
      </div>
    </div>
  </div>
</form>
`;
